tracer/src/Datadog.Tracer.Native/clr_helpers.cpp

```diff
@@ -1018,7 +1018,318 @@ HRESULT HasAsyncStateMachineAttribute(const ComPtr<IMetaDataImport2>& metadataIm
+                        if (assemblyRefMetadata.name == shared::EmptyWStr)
+                        {
+                            Logger::Warn(
+                                "[ResolveTypeInternal] GetResolutionScopeNameForAssemblyRefTok has failed with: ", shared::WSTRING(refTypeName.data()));
+                            continue;
+                        }
+
+                        hr = ResolveTypeInternal(info, loadedModules, refTypeName, mdExportedTypeNil, assemblyRefMetadata.name,
+                                                resolvedTypeDefToken, resolvedTypeDefMetadataImport);
+                        if (FAILED(hr))
+                        {
+                            Logger::Warn(
+                                "[ResolveTypeInternal] Recursive call to ResolveTypeInternal has failed with: ", shared::WSTRING(refTypeName.data()));
+                            continue;
+                        }
+
+                        return hr;
+                    }
+                }
+            }
+        }
+    }
+
+    if (isModuleFound)
+    {
+        return S_OK;
+    }
+    
+    resolvedTypeDefToken = mdTokenNil;
+    Logger::Error("[ResolveTypeInternal] ResolveTypeInternal has failed. Reason: Module not found in the loaded modules for type name: ", shared::WSTRING(refTypeName.data()));
+    return E_NOTIMPL;
+}
+
+/**
+ Resolves TypeRef to TypeDef.
+ Note: If the module where the TypeRef is defined is not loaded, E_FAIL will return.
+ */
+HRESULT ResolveType(ICorProfilerInfo4* info,
+                    const ComPtr<IMetaDataImport2>& metadata_import,
+                    const ComPtr<IMetaDataAssemblyImport>& assembly_import,
+                    const mdTypeRef typeRefToken,
+                    mdTypeDef& resolvedTypeDefToken,
+                    ComPtr<IMetaDataImport2>& resolvedMetadataImport)
+{
+    mdToken resolutionScope = mdTokenNil; // will hold either AssemblyRef or ModuleRef token
+    mdToken enclosingType = mdTokenNil;
+    ULONG nameSize = 0;
+
+    auto hr = metadata_import->GetTypeRefProps(typeRefToken, &resolutionScope, nullptr, 0, &nameSize);
+    if (FAILED(hr) || resolutionScope == mdTokenNil)
+    {
+        Logger::Error("[ResolveType] GetTypeRefProps [1] has failed. typeRefToken: ", typeRefToken);
+        return E_FAIL;
+    }
+    // get the type name & resolution scope
+    std::vector<WCHAR> refTypeName(nameSize, 0);
+    hr = metadata_import->GetTypeRefProps(typeRefToken, &resolutionScope, refTypeName.data(),
+                                                          kNameMaxSize, &nameSize);
+    if (FAILED(hr) || resolutionScope == mdTokenNil)
+    {
+        Logger::Error("[ResolveType] GetTypeRefProps [2] has failed. typeRefToken: ", typeRefToken);
+    }
+    IfFailRet(hr);
+    mdToken tempToken = mdTokenNil;
+    // To avoid ending up in an infinite loop, I'm limiting the execution to 1000 (arbitrary large number that will be
+    // well beyond enough)
+    int tryoutsCount = 1000;
```

Comment:
retryCount

Reply:
Done in ef502336ac5073c545e222bc6396635a9d76288f.